### ISSUE TITLE
Erratic frame pacing after 5-30 seconds (applies to both Master and Dev builds)

### ISSUE THREAD

**Issue by RavenMacDaddy** [flyinghead/flycast#732]
Platform / OS / Hardware: 

Windows 11 - 22000.918 (latest at the time of writing)

GitHub hash: 

[aa97a6d](https://github.com/flyinghead/flycast/commit/aa97a6d) (Master)
[e0d87eb](https://github.com/flyinghead/flycast/commit/e0d87eb) (Dev)

Latest at the time of writing.

Hardware: 

AMD Ryzen 3600 @ 4.2 GHz
32 GB DDR4 3000 MHz RAM
NVIDIA GTX 1080

**_Description of the Issue_**

When running with the Vulkan backend, after about 5-30 seconds from starting a game, the frame pacing becomes erratic - as can be seen both in terms of the consistency and smoothness of the contents on-screen, but also the OSD of the monitor which provides an FPS counter.

The monitor is adjusting the refresh rate to the contents thanks to using G-Sync.

...and yes, I'm running G-Sync with multiple other apps/games without issues; it's not the G-Sync feature that's malfunctioning.

With the DX11 backend, it's always erratic, from the moment you start a game.

The OpenGL backend is not feasible right now as running for example a NAOMI 2 GD-ROM game like Virtua Fighter 4: Final Tuned, the GPU gets completely tanked if I run with my preferred setting of 6x internal res; 2880p/5K - the ones above perform just fine in comparison performance wise.

If I switch between games, the behavior repeats (with the Vulkan backend); at first, the frame pacing is fine, but after 5-30 seconds, like clockwork, it starts to go all over the place.

I've tested the following games:

Rayman 2 (Dreamcast)
Power Stone 2 (NAOMI)
Maze of the Kings (NAOMI GD-ROM)
Virtua Fighter 4: Final Tuned (rev B) (NAOMI 2 GD-ROM)

...after which point, it was clear that the platform doesn't matter - it's seemingly an overall issue.

**_Debugging Steps Tested_**

* Tested both the latest Master and Dev builds
* Tested different games on different platforms
* Testing different APIs
* Running with native res rather than enhanced

 
**_Logs Gathered_**

Didn't find any logs in the emulator dir - if needed, just ask and tell me how.

**_Screenshots_**

https://user-images.githubusercontent.com/67732686/188281759-db58395b-e861-4779-8407-0772f637b6a5.mp4

**Comment by flyinghead** [comment]
I'm afraid there's some misunderstanding with how "adaptive sync" technologies work with emulators and Flycast in particular. Adaptive sync makes sense with a game that can output, say, 120 FPS most of the time, with lower frame rate during heavy scenes. Then the GPU can sense the current frame rate and sync the monitor refresh rate in real time.

However, (if enabled) Flycast uses the monitor VSync as a synchronization source with "real time", the only other source being the audio output. Unfortunately these sources aren't perfect: the audio output buffer won't block until it's full, which depends on the platform and audio driver used. So using VSync as well is highly recommended. With a 120+ Hz monitor, Flycast will skip one or more VSync each frame to maintain 60 FPS (but not with Vulkan because this feature isn't available: you need to enable the `Duplicate Frame` option).

In other words, adaptive sync with flycast is like your gpu syncing to flycast, syncing to your gpu, syncing to flycast, ...
There's no way out

If you can, force Flycast to a strict 60 FPS and you should be good.

**Comment by RavenMacDaddy** [comment]
That's the thing - it's capped at 60.

For reference, emulators like FinalBurn Neo with the same cap/situation (RTSS, which also fixes frame times on that end) stays locked at 120 Hz for the majority of the time - a feature that's called Low Frame rate Compensation, or LFC for short. 

LFC multiplies the refresh rate of the monitor in relation to the FPS of what's rendered, and since my monitor's max refresh rate is 175 Hz, 60 FPS equals 120 Hz when the frame rate is stable. 

That's why the OSD says 120 Hz at first; the content is first running at a solid 60 FPS before it starts glitching out.

**Comment by flyinghead** [comment]
Your assumption is that Flycast will/should render at a stable 60 FPS if uncapped. This is NOT the case.

This is true when averaged over some period of time but not between 2 frames. Here is what happens on a 120 Hz monitor (with vulkan, swap interval = 1, no frame dupe).
```
Screen:
* * * * * * *
Expected behavior:
*   *   *   *
Actual behavior:
* *     * *
```
In other words, if given the opportunity, Flycast will sometimes render faster than 60 FPS for a couple of frames. Then it will block on audio for some time without rendering. Frame time in this case is obviously all over the place.

Except with Vulkan, when VSync is enabled, Flycast will use a swap interval greater than 1 if it detects a 120+ Hz monitor. So it will skip every other VSync and thus you get the expected behavior above. For Vulkan, you need to enable Duplicate Frame for the same effect.

Don't use variable refresh rate/LFC with Flycast. Use VSync or RTSS to force it to render at 60 FPS.

**Comment by RavenMacDaddy** [comment]
RTSS is always being used to cap the frame rate, and for a few seconds, the presentation is perfect.

The frame times are consistent (when Vulkan is being used), which means G-Sync and the monitor's refresh rate is locked to 120 Hz for the majority of the time.

Then something happens that leads to the frame times beginning to jump – and judder is the result of that.

To be clear, this all happens whether G-Sync is being used **or not** – its usage simply not a factor, but it's what makes the issue clear (thanks to the monitor's OSD).

It's also what makes the presentation super smooth when everything is working as intended.

---

If it'd help, I'll gladly provide logs if you tell me how.

I appreciate your work, and I also appreciate how difficult it must be to figure something like this out.

I don't blame you if you decide that this issue is not worth putting into – but the potential is there; the presentation starts out perfectly, and it looks so smooth and consistent during that time.

**Comment by rapego9596** [comment]
Yes. Vulkan runs like shit with frame pacing issues all over the place. This is similiar to the bug I posted here with video for reference:

https://github.com/flyinghead/flycast/issues/699

It seems like the dev doesn't want to acknowledge the issue. What makes it more ridiculous is using all settings maxed with ample GPU overhead, he blamed it on my hardware "on a Mac" (like the M1 Pro is a POS or something). Forced to use OpenGL indefinitely or until he cares one day to really look into it.

**Comment by RavenMacDaddy** [comment]
@rapego9596 

I'd rather give the developer the benefit of the doubt - and as described this isn't isolated to the Vulkan API, which makes me think the issue is rather in the emulator as a whole; Vulkan only fairs better for a while. 

Refrain from bringing the tone down - keep the discussion objective.

**Comment by rapego9596** [comment]
There's constructive, which already happened. There is also ignoring a problem, which also happened. Bottom line, it's an ongoing issue with Vulkan. OpenGL runs fine for me, but loses accuracy due to lack of per-pixel option.

**Comment by agarcia8792** [comment]
Hello guys,

I Just wanted to report that I finally found a fix to this problem. Almost a year late, but I see that many people still have this problem.

First, sadly the solutions proposed by the dev here don't seem to fix the issue, at least for me. Vsync on vs off, VRR on vs off, 60fps cap, nothing seems to work.

So, the solution I found is to cap the framerate to exactly 59.93 with Riva Tuner in combination with the emulator's 59.94hz fixed frequency option (or "auto", in case you selected VGA). Vsync seems irrelevant and I'm also using a 120hz Gsync monitor.

This will give sustained 59 FPS with no spikes/dips after periods of time.

This, however isn't a perfect solution. In games like MVC2 it is necessary to pause/unpause the emulator every time a fight starts, since desyncing occurs in the menus or the character select screen. Other games like MVC1 maintain the framerate permanen

**Comment by agarcia8792** [comment]
Hello guys,

I just wanted to report that I finally found a fix to this problem. I know I'm a little bit late, but I see that many people still have this problem even with the latest versions. I'm using Flycast Dojo btw. It gives the best results since it has the fixed frequency options which are important.

First, sadly the solutions proposed by the dev here don't seem to work, at least for me. Vsync on vs off, VRR on vs off, 60fps cap, nothing works.

**The solution I found is to cap the framerate at exactly 59.93 with RTSS in conjunction with the emulator's 59.94hz fixed frequency option (or "auto", in case you're using NTSC/VGA).** Vsync is irrelevant and I'm also using a +120 Hz VRR monitor.

This will give permanent 59/60 FPS with no spikes/dips regardless of how much gameplay time has passed. Also, no audio glitches that would occur if the framerate is capped below 59.93 (that's why 59.93 is the magic number). If you ask, double or single capping at 59.94 or higher also results in failure.

This however, isn't a perfect solution. In games like MVC2 (Dreamcast) it is necessary to pause/resume the emulator (not the game) every time a fight starts, since some desyncing issues happen during transitions and character select screens (Vsync does not help). Due to this, training mode works perfectly well after you pause/resume the emulator a single time while in the fight. This pause/resume thing was already a workaround I discovered, but as the person who opened this thread stated, the effect lasts around of 30 seconds before everything goes crazy. Now I've made it permanent, but at least in the case of MVC2, it is necessary to stay in fight. Other games like MVC1 are able to maintain the framerate permanently throughout the entire gameplay (a one time pause/resume is still required, but it doesn't need to happen in a specific moment).

These are other relevant settings I used to test this:
DirectX 11
Fixed Freq CPU optimizations = OFF
Multi-Threaded Emulation = ON
Audio Frame Sync = ON

My PC specs

Core i7 9700K
RTX 3080
165 Hz Gsync compatible monitor capped at 120 Hz (DisplayPort 1.4 connection)
Windows 11

Sorry for any grammar mistake. English is not my first language.

**Comment by CookiePLMonster** [comment]
Bumping this instead of creating a new topic to signal that:

1. This is still an issue in v2.7.
2. It impacts non-VRR environments running at constant 60Hz.

I observed erratic frame times (despite maintaining a stable average 60FPS) even in the bios menu, but these clips from V-Rally 2 make it extremely evident that something is wrong. Even if you don't see the difference between the two clips at a glance, downloading both and stepping frame-by-frame makes it obvious.

This clip, using OpenGL, has spotless frame pacing. Every frame is unique, and the car advances steadily:
https://github.com/user-attachments/assets/9608b6c7-23f2-473d-9a6b-73cfef4c419b

Using Vulkan, frame pacing is messy even though the framerate is stable. The car advances steadily for two frames, then two next frames are duplicated, then there's a bigger jump, and so on:
https://github.com/user-attachments/assets/166db300-7284-48fa-81f6-bbacce2e0dcf

**Comment by flyinghead** [comment]
@CookiePLMonster I pushed a fix that may be the cause of frame losses in Vulkan. This fix only applies to NVidia GPUs.
EDIT: this is on the dev branch

**Comment by CookiePLMonster** [comment]
Thanks, I'll give it a try and report back! A few questions:

1. In this case, was OGL always using FIFO? I assume so, as iirc Mailbox is VK only (Fast Sync isn't readily available elsewhere).
2. Have you tried FIFO Relaxed? With a hard deadline of 60Hz, a vblank-or-tear behaviour that Relaxed has may be preferable, but IIRC hardware support varies. Around 2 years ago I was unable to leverage FIFO Relaxed on a GTX 10 series properly, at least. Now I'm on a 50 series, so it may have changed.
3. With Mailbox/Vsync on, does Flycast try to present at fixed 60Hz intervals? The current frame times kind of indicate that it doesn't, which would explain why Mailbox behaved so badly.

**Comment by flyinghead** [comment]
Mailbox is a vulkan things so other APIs are using FIFO.
FIFO relaxed produced bad frame pacing in my (quick) test so I didn't bother.
With mailbox, presenting a frame too quickly will overwrite the previous one waiting to be presented, resulting in frame loss. But I haven't been able to confirm it.

**Comment by CookiePLMonster** [comment]
> Mailbox is a vulkan things so other APIs are using FIFO.

This explains the disparity then, and indeed, using FIFO produces good results. Didn't even need to look at any frame time graphs to see na immediate improvement:

https://github.com/user-attachments/assets/17bb5fda-9f97-4f73-8e69-89878e2d7809

> With mailbox, presenting a frame too quickly will overwrite the previous one waiting to be presented, resulting in frame loss. But I haven't been able to confirm it.

My suspicion is that with Mailbox, maybe the the presentation throttling done by audio ended up being out of sync with the display's vblank and ended up submitting frames just after the vblank (hence those strange 2ms-ish frame times). But with FIFO, the blocking behaviour makes it impossible to desync as badly.
